### ISSUE TITLE
Verification of user's group fails in LDAP backend.

### ISSUE THREAD

**Issue by lbogardi** [StackStorm/st2#2721]
Hi Folks, @m4dcoder 

after setting up LDAP auth backend, I found, that ldap binding works fine, it finds the user to be authenticated, but, when it comes to the group validation, it fails. After some debugging,
I get the following result:

`2016-05-30 17:09:24,507 41870672 INFO app [-] st2auth.app app created.
2016-05-30 17:09:39,828 56682928 ERROR ldap_backend [-] Query"(|(&(objectClass=*)(|(member=uid=laszlo,ou=people,dc=office,dc=ams-ix,dc=net)(uniqueMember=uid=laszlo,ou=people,dc=office,dc=ams-ix,dc=net))))"
None
2016-05-30 17:09:39,828 56682928 ERROR ldap_backend [-] Unable to verify membership for user "laszlo".
None
2016-05-30 17:09:39,829 56682928 AUDIT auth [-] Invalid credentials provided (remote_addr='127.0.0.1',auth_backend='LDAPAuthenticationBackend')`
The "Query" in the output is added by me.

However, the following ldapsearch command works fine:
`ldapsearch -H ldaps://ldap.ams-ix.net -x -b ou=groups,dc=office,dc=ams-ix,dc=net -D uid=st2_auth,ou=applications,dc=office,dc=ams-ix,dc=net -W '(cn=noc)'`

and it returns the members of 'noc' group.

The related part of st2.conf is:
`backend_kwargs = {"bind_dn": "uid=st2_auth,ou=applications,dc=office,dc=ams-ix,dc=net", "bind_password": "********", "base_ou": "dc=office,dc=ams-ix,dc=net", "group_dns": ["cn=noc,ou=groups,dc=office,dc=ams-ix,dc=net"], "host": "ldap.ams-ix.net", "port": 636, "use_ssl": true, "cacert": "/etc/ssl/st2/ldap.crt"}`

Additionally, if I comment out the related try - except block from this:
`/opt/stackstorm/st2/lib/python2.7/site-packages/st2auth_enterprise_ldap_backend/ldap_backend.py` file, everything works fine, and the user gets authenticated.

Please let me know, if you need any more information.
TIA,
Cheers

**Comment by Kami** [comment]
Thanks for the report.

I'm not exactly sure why it's not working (it looks like a invalid query, but I would need to dig in), but @m4dcoder might have more clue :)

For one, we need to add more debug logging to the backend so it will make things easier to debug:

1. Log actual connection URL (that's important to see if SSL / TLS is used)
2. Log information which was retrieved by the server (user groups, etc.)

On a related note, @m4dcoder does it make sense to make membership check optional (the thing which is causing issues here)? I'm not too sure, since I think it's usually a good idea to verify that user is part of some specific group which should have access to StackStorm.

**Comment by lbogardi** [comment]
Hey @Kami,

here is what you've asked for:

`2016-05-31 14:08:10,965 41907536 INFO __init__ [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2016-05-31 14:08:11,020 41907536 INFO app [-] st2auth.app app created.
2016-05-31 14:08:19,620 56826288 ERROR ldap_backend [-] Endpoint: "ldaps://ldap.ams-ix.net:636"
None
2016-05-31 14:08:19,620 56826288 ERROR ldap_backend [-] Group DNs: "[u'CN=noc,ou=groups,dc=office,dc=ams-ix,dc=net']"
None
2016-05-31 14:08:19,621 56826288 ERROR ldap_backend [-] Executed Query: "(|(&(objectClass=*)(|(member=uid=laszlo,ou=people,dc=office,dc=ams-ix,dc=net)(uniqueMember=uid=laszlo,ou=people,dc=office,dc=ams-ix,dc=net))))"
None
2016-05-31 14:08:19,621 56826288 ERROR ldap_backend [-] Result of Query: "[]"
None
2016-05-31 14:08:19,621 56826288 ERROR ldap_backend [-] Unable to verify membership for user "laszlo".
None
2016-05-31 14:08:19,622 56826288 AUDIT auth [-] Invalid credentials provided (remote_addr='127.0.0.1',auth_backend='LDAPAuthenticationBackend')
`

And the code which produced this:
( earlier in the code, I stored endpoint into self, so I can print it later )

`          
        # Setup connection and options.
        protocol = 'ldaps' if self._use_ssl else 'ldap'
        endpoint = '%s://%s:%d' % (protocol, self._host, int(self._port))

        # store endpoint in self
        self._endpoint = endpoint

        ...

            # Search if user is member of pre-defined groups.
            try:
                query = '(|(&(objectClass=*)(|(member={0})(uniqueMember={0}))))'.format(user_dn)
                result = connection.search_s(self._base_ou, self._scope, query, [])
                entries = [entry[0] for entry in result if entry[0] is not None] if result else []

                if not list(set(self._group_dns) & set(entries)):
                    LOG.exception('Endpoint: "%s"' % self._endpoint )
                    LOG.exception('Group DNs: "%s"' % self._group_dns )
                    LOG.exception('Executed Query: "%s"' % query )
                    LOG.exception('Result of Query: "%s"' % result )
                    LOG.exception('Unable to verify membership for user "%s".' % username)
                    return False
            except Exception:
                LOG.exception('Unexpected error when querying membership for user "%s".' % username)
                return False
`

Or, if you prefer, just give me some code I should execute with some debug statements.
I think this is my first python code, I ever touched. :)

Cheers

**Comment by sbakker** [comment]
Sorry for butting in.

I've fixed this before for CheckMk, see http://lists.mathias-kettner.de/pipermail/omd-devel/2013-July/000450.html.

It looks like the code is expecting group membership to be implemented in objects with parent classes `groupOfNames` or `groupOfUniqueNames`, using attributes `member` or `uniqueMember`. We use `posixGroup` with the `memberUid` attribute. The `memberUid` attribute should contain only the `username` (`uid` attribute), not a complete DN.

In this case, I think the fix would be:

````
query = '(|(&(objectClass=*)(|(member={0})(uniqueMember={0})(memberUid={1})))'.format(user_dn, username)
````

**Comment by Kami** [comment]
@lbogardi Thanks for providing additional details.

@sbakker Thanks for the pointers - we will look into it (and also amend and improve our tests :)).

/cc @m4dcoder

**Comment by lbogardi** [comment]
Hi @m4dcoder, @Kami 

a possible (working in our case) solution for the problem:

```
            # Search if user is member of pre-defined groups.
            try:
                query = '(|(&(objectClass=*)(|(memberUid={0})(uniqueMember={0}))))'.format(username)
                result = connection.search_s(self._base_ou, self._scope, query, [])

                entries = [entry[0] for entry in result if entry[0] is not None] if result else []
                members = entry[1]['memberUid']

                if not list(set(members) & set([username])):
                    LOG.exception('Unable to verify membership for user "%s".' % username)
                    return False
            except Exception:
                LOG.exception('Unexpected error when querying membership for user "%s".' % username)
                return False
```

as it was pointed out by @sbakker, the uniqueMember filter might be unneccessary, too.

Hope it helps,
Cheers

**Comment by codyaray** [comment]
Fwiw, I'm in favor of not _requiring_ but _supporting_ group membership restriction. When I first got started with St2 Enterprise, we didn't yet have an A/D group for it created. Since I had to specify a group, I ended up just piggybacking on another group for a while.

PS hi @sbakker! surprised to see a fellow check_mk user here! I've got a super-basic [check_mk PR](https://github.com/StackStorm/st2contrib/pull/475) pending right now if you're interested. :)
